stargate-blog/tutorial/02-cli-list-post.md

```diff
@@ -0,0 +1,357 @@
+      return getPost(ctx, path[1], k, legacyQuerierCdc)
+
+    case types.QueryListPost:
+      return listPost(ctx, k, legacyQuerierCdc)
+
+    default:
+```
+
+Now let’s define `listPost`:
+
+### x/blog/keeper/query_post.go
+
+Make sure to add the codec to the previous import and add the `listPost` and `getPost` function in a new file called `query_post.go` in our keeper.
+
+```go
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+func listPost(ctx sdk.Context, keeper Keeper, legacyQuerierCdc *codec.LegacyAmino) ([]byte, error) {
+	msgs := keeper.GetAllPost(ctx)
+
+	bz, err := codec.MarshalJSONIndent(legacyQuerierCdc, msgs)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONMarshal, err.Error())
+	}
+
+	return bz, nil
+}
+
+func getPost(ctx sdk.Context, id string, keeper Keeper, legacyQuerierCdc *codec.LegacyAmino) ([]byte, error) {
+	msg := keeper.GetPost(ctx, id)
+
+	bz, err := codec.MarshalJSONIndent(legacyQuerierCdc, msg)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONMarshal, err.Error())
+	}
+
+	return bz, nil
+}
+
+```
+
+In the keeper we define also the grpc of our queryPost function.
+
+### x/block/keeper/grpc_query_post.go
+
+```go
+package keeper
+
+import (
+	"context"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/example/blog/x/blog/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) PostAll(c context.Context, req *types.QueryAllPostRequest) (*types.QueryAllPostResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	var posts []*types.Post
+	ctx := sdk.UnwrapSDKContext(c)
+
+	store := ctx.KVStore(k.storeKey)
+	postStore := prefix.NewStore(store, types.KeyPrefix(types.PostKey))
+
+	pageRes, err := query.Paginate(postStore, req.Pagination, func(key []byte, value []byte) error {
+		var post types.Post
+		if err := k.cdc.UnmarshalBinaryBare(value, &post); err != nil {
+			return err
+		}
+
+		posts = append(posts, &post)
+		return nil
+	})
+
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &types.QueryAllPostResponse{Post: posts, Pagination: pageRes}, nil
+}
+
+func (k Keeper) Post(c context.Context, req *types.QueryGetPostRequest) (*types.QueryGetPostResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	var post types.Post
+	ctx := sdk.UnwrapSDKContext(c)
+
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.PostKey))
+	k.cdc.MustUnmarshalBinaryBare(store.Get(types.KeyPrefix(types.PostKey + req.Id)), &post)
+
+	return &types.QueryGetPostResponse{Post: &post}, nil
+}
+
+```
+
+We add the grpc query handler to our module on 
+
+### x/blog/module.go
```

Comment:
I had an error that `context` is undefined, so we probably need to add a snippet like "make sure to import `context`".